### DISCUSSION TITLE
Understanding the relationship of tagPrefix and suffix

### DISCUSSION THREAD

**Discussion by jack1902** [strimzi/strimzi-kafka-operator#10565] [Q&A]
I'm trying to understand the tradeoffs of using `image.tagPrefix` vs `image.tag`. 

I see the helm-chart values.yaml has `tagPrefix` listed here

https://github.com/strimzi/strimzi-kafka-operator/blob/8e30be30de02118ab43e12f45304039419d55ff3/helm-charts/helm3/strimzi-kafka-operator/values.yaml#L99

But the [README.md](https://github.com/strimzi/strimzi-kafka-operator/blob/main/helm-charts/helm3/strimzi-kafka-operator/README.md) has the below

![image](https://github.com/user-attachments/assets/b73c7ef2-5b15-42d4-8a1d-a1ecf1ebfe27)

I am struggling to find documentation on the tradeoffs of using `tagPrefix` vs `tag`. If i use `tag`, how does that impact upgrades of kafka, if the downstream `connect` CR has `.spec.version: some-future-version` does that result in it using the `tag` defined on the operator regardless?

Where does the `suffix` come from?

**Comment by scholzj** [comment]
If you specify some unsupported Kafka version, you should always get an error. So it does not matter what image you have configured.

the main point of the `tagPrefix` is that many images have the tags consist of several parts - e.g. `quay.io/strimzi/kafka:0.43.0-kafka-3.7.0`. And the `tagPrefix` let's you configure the first part of the tag: `<tagPrefix>-kafka-3.7.0`. The `tag` on the other hand should configure the whole tag IIRC.

**Reply by jack1902** [reply]
Makes sense, although it seems that if i use `tagPrefix` it adds the `kafka-VERSION` bit. If i use `tag` only, then it basically means the operator supports `3.6,3.7,3.8`, the image that would be pulled will always be the `tag` i have defined (this isn't an issue as i tend to produce the newer images first before bumping the kafkaVersion in use).

That is the tradeoff it seems, whilst the operator itself can support multiple versions of the `.spec.version` to create the `kafka-....` suffix, if i have defined `tag` only, then i only support the version within that tag (although if i have a tag with kafka-3.8.0, technically `3.6,3.7` would still work? [although not 100% sure there])

**Reply by scholzj** [reply]
The Kafka version supported is not only about the images. There is no Strimzi version that supports 3.6, 3.7, and 3.8. But yes, the tag overrides the whole tag and you will end up with the same tag for all supported versions. That is why it makes little sense for the versioned images. It makes more sense for non-versioned images.

**Reply by jack1902** [reply]
Apologies if i caused confusion on the `3.6,3.7,3.8` piece, was trying to lay down an example.

![image](https://github.com/user-attachments/assets/8035ed81-0035-49a5-842f-26a588868c21)

Viewing the above from https://strimzi.io/downloads/ i see the operator `0.43.0` supports `3.7.0,3.7.1,3.8.0`. If i set the `image.tag` (specifically for the kafka-connect) to something such as `3.8.0`, but i had set my spec.version within the connect CR to use `3.7.0` would there be any fallout?

I'm really trying to ensure that i don't cause any problems by using `image.tag` and the relationship between the suffix. Its clearly supported to use `image.tag` instead of `image.tagPrefix` but i don't yet understand the fallout.

**Reply by jack1902** [reply]
I understand the `suffix` won't be used if i use `image.tag`, but i'm attempting to understand the caveats of using `image.tag` over `image.tagPrefix`

**Reply by scholzj** [reply]
>  If i set the image.tag (specifically for the kafka-connect) to something such as 3.8.0, but i had set my spec.version within the connect CR to use 3.7.0 would there be any fallout?

The image for Kafka version X should always contain Kafka version X. Otherwise you might run into issues. Connect does not have too many changes. But it can happen that for example some new API call is added in 3.8 that is missing in 3.7. And if you tell the operator to use Kafka 3.8, but give it an image of Kafka 3.7 instead, it would not work when the operator tries to use the new API. So the images should always correspond to the given version.

As I said, configuring the `tag` is not really useful for the images based on a particular Kafka version. But it is useful for example to configure the UserOperator image as there you have only one image for all Kafka versions.

**Reply by jack1902** [reply]
Cheers, will ensure that i have internal images for all possible suffixes within a given release. In the case of `0.43.0` above, the connect image suffix of `kafka-3.8.0` comes from the `.spec.version` of the connect CR right?

Based on that, i would need to ensure i have `kafka-3.7.0`, `kafka-3.7.1` and `kafka-3.8.0` available internally? (what drives the suffix?)

**Reply by scholzj** [reply]
In general yes. Of course, the image will be only used if you use given version. So if you know you are only using 3.7.0 and 3.8.0, the 3.7.1 might never be used and while you might need to set it to some value, the actual image might not need to exist. But in general it is easiest to just have all the images available.

**Reply by jack1902** [reply]
Cheers, thank you for your patience here by the way, will close this out :bow:
